File: internal/server/frames.go
package server

import "github.com/go-faster/errors"

// Frame types multiplex the pty↔browser websocket, which otherwise carries
// only opaque bytes: a single binary websocket message can be terminal
// output, terminal input, or a resize instruction, so every message on the
// /ws stream is tagged with a one-byte type prefix.
const (
	// FrameOutput carries raw pty stdout/stderr bytes, server → client.
	FrameOutput byte = 'o'
	// FrameInput carries raw keystroke/paste bytes, client → server.
	FrameInput byte = 'i'
	// FrameResize carries a 4-byte cols/rows payload (see EncodeResize),
	// client → server, sent whenever the browser's fit-addon recomputes
	// the terminal size.
	FrameResize byte = 'r'
	// FrameError carries a human-readable error message, server → client,
	// sent when the pty/herdr spawn fails or herdr becomes unreachable so
	// the frontend can show it instead of hanging on a dead connection.
	FrameError byte = 'e'
)

// EncodeFrame prepends typ to data, producing the wire format for one
// websocket binary message on the /ws stream. Pure and allocation-cheap
// (one prepend) since it runs on every pty read in the hot path.
func EncodeFrame(typ byte, data []byte) []byte {
	out := make([]byte, len(data)+1)
	out[0] = typ
	copy(out[1:], data)
	return out
}

// DecodeFrame splits a raw websocket binary message back into its type tag
// and payload. Errors on an empty message, which is never valid on this
// stream (EncodeFrame always emits at least the type byte).
func DecodeFrame(msg []byte) (typ byte, data []byte, err error) {
	if len(msg) == 0 {
		return 0, nil, errors.New("empty frame")
	}
	return msg[0], msg[1:], nil
}

// EncodeResize packs a terminal size into a FrameResize payload: cols and
// rows as big-endian uint16s. uint16 comfortably covers any real terminal
// (max 65535 cols/rows) while keeping the resize frame fixed-size and
// alloc-free to decode.
func EncodeResize(cols, rows uint16) []byte {
	return []byte{byte(cols >> 8), byte(cols), byte(rows >> 8), byte(rows)}
}

// DecodeResize is EncodeResize's inverse. Errors if data isn't exactly 4
// bytes, which means the client sent a malformed resize frame.
func DecodeResize(data []byte) (cols, rows uint16, err error) {
	if len(data) != 4 {
		return 0, 0, errors.Errorf("resize frame must be 4 bytes, got %d", len(data))
	}
	cols = uint16(data[0])<<8 | uint16(data[1])
	rows = uint16(data[2])<<8 | uint16(data[3])
	return cols, rows, nil
}

File: frontend/src/lib/frames.ts
/**
 * Wire format for the /ws pty↔browser stream. Mirrors
 * internal/server/frames.go exactly — a change here must be mirrored there
 * (and vice versa) or the two sides silently stop understanding each
 * other's frames.
 *
 * Every websocket binary message is `[type byte][payload...]`. This module
 * has zero DOM/xterm dependency so it can be unit tested (and, later,
 * reused by ticket 6's transport) without a browser.
 */

export const FRAME_OUTPUT = 0x6f // 'o' — pty stdout/stderr, server -> client
export const FRAME_INPUT = 0x69 // 'i' — keystrokes/paste, client -> server
export const FRAME_RESIZE = 0x72 // 'r' — 4-byte cols/rows, client -> server
export const FRAME_ERROR = 0x65 // 'e' — human-readable error, server -> client

export type FrameType =
  | typeof FRAME_OUTPUT
  | typeof FRAME_INPUT
  | typeof FRAME_RESIZE
  | typeof FRAME_ERROR

/** Prepends a type byte to data. Pure; used on every keystroke and resize. */
export function encodeFrame(type: FrameType, data: Uint8Array): Uint8Array {
  const out = new Uint8Array(data.length + 1)
  out[0] = type
  out.set(data, 1)
  return out
}

/** Splits a raw websocket message back into its type byte and payload. */
export function decodeFrame(msg: Uint8Array): { type: number; data: Uint8Array } {
  if (msg.length === 0) {
    throw new Error('empty frame')
  }
  return { type: msg[0], data: msg.subarray(1) }
}

/** Packs a terminal size as the FRAME_RESIZE payload: cols, rows as big-endian uint16s. */
export function encodeResize(cols: number, rows: number): Uint8Array {
  return new Uint8Array([
    (cols >> 8) & 0xff,
    cols & 0xff,
    (rows >> 8) & 0xff,
    rows & 0xff,
  ])
}

const textEncoder = new TextEncoder()
const textDecoder = new TextDecoder()

/** Encodes a JS string as an input frame ready to send over the ws. */
export function encodeInput(text: string): Uint8Array {
  return encodeFrame(FRAME_INPUT, textEncoder.encode(text))
}

/** Encodes a resize as a ready-to-send frame. */
export function encodeResizeFrame(cols: number, rows: number): Uint8Array {
  return encodeFrame(FRAME_RESIZE, encodeResize(cols, rows))
}

/** Decodes an error frame's payload back to a human-readable string. */
export function decodeErrorMessage(data: Uint8Array): string {
  return textDecoder.decode(data)
}
